<a href="https://colab.research.google.com/github/Balajilp/PySpark-Learning/blob/main/Datagrokr_Assessment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **SECTION 1**



In [ ]:
!apt-get install openjdk-8-jdk-headless -qq > /dev/null

In [ ]:
!wget -q https://archive.apache.org/dist/spark/spark-3.1.2/spark-3.1.2-bin-hadoop2.7.tgz

In [ ]:
!tar xf spark-3.1.2-bin-hadoop2.7.tgz

In [ ]:
import os
os.environ["JAVA_HOME"] = "/usr/lib/jvm/java-8-openjdk-amd64"

os.environ["SPARK_HOME"] = "/content/spark-3.1.2-bin-hadoop2.7"

In [ ]:
!pip install -q findspark
import findspark
findspark.init()

In [ ]:
# integrating google colab with drive
from google.colab import drive
drive.mount('/content/drive/')

Drive already mounted at /content/drive/; to attempt to forcibly remount, call drive.mount("/content/drive/", force_remount=True).


In [ ]:
# Creating sparksession
from pyspark.sql import SparkSession
spark = (SparkSession
.builder
.appName("My_work")
.getOrCreate())

In [ ]:
# Reading the Venue Csv file
df_venue = spark.read.csv('/content/drive/MyDrive/Colab Notebooks/DataSet/ipl_venue.csv', header=True, inferSchema=True)

# printing the Schema
df_venue.printSchema()

root
 |-- venue_id: integer (nullable = true)
 |-- venue: string (nullable = true)
 |-- city: string (nullable = true)



In [ ]:
# printing the top five records
df_venue.show(5)

+--------+--------------------+-------------+
|venue_id|               venue|         city|
+--------+--------------------+-------------+
|       1|Dr. Y.S. Rajasekh...|Visakhapatnam|
|       2|Sharjah Cricket S...|      Sharjah|
|       3|JSCA Internationa...|       Ranchi|
|       4|Saurashtra Cricke...|       Rajkot|
|       5|Shaheed Veer Nara...|       Raipur|
+--------+--------------------+-------------+
only showing top 5 rows



In [ ]:
# Reading the Ball by Ball csv file
df_ipl_ball = spark.read.csv('/content/drive/MyDrive/Colab Notebooks/DataSet/ipl_ball_by_ball.csv', header=True, inferSchema=True)

# printing Schema
df_ipl_ball.printSchema()

root
 |-- match_id: integer (nullable = true)
 |-- inning: integer (nullable = true)
 |-- overs: integer (nullable = true)
 |-- ball: integer (nullable = true)
 |-- batsman: string (nullable = true)
 |-- non_striker: string (nullable = true)
 |-- bowler: string (nullable = true)
 |-- batsman_runs: integer (nullable = true)
 |-- extra_runs: integer (nullable = true)
 |-- total_runs: integer (nullable = true)
 |-- non_boundary: integer (nullable = true)
 |-- is_wicket: integer (nullable = true)
 |-- dismissal_kind: string (nullable = true)
 |-- player_dismissed: string (nullable = true)
 |-- fielder: string (nullable = true)
 |-- extras_type: string (nullable = true)
 |-- batting_team: string (nullable = true)
 |-- bowling_team: string (nullable = true)



In [ ]:
# printing the top 5 records
df_ipl_ball.show(5)

+--------+------+-----+----+-------------+-------------+------------+------------+----------+----------+------------+---------+--------------+----------------+-------+-----------+--------------------+--------------+
|match_id|inning|overs|ball|      batsman|  non_striker|      bowler|batsman_runs|extra_runs|total_runs|non_boundary|is_wicket|dismissal_kind|player_dismissed|fielder|extras_type|        batting_team|  bowling_team|
+--------+------+-----+----+-------------+-------------+------------+------------+----------+----------+------------+---------+--------------+----------------+-------+-----------+--------------------+--------------+
|  419157|     2|   16|   4|      V Kohli|R Vinay Kumar|CRD Fernando|           0|         0|         0|           0|        0|            NA|              NA|     NA|         NA|Royal Challengers...|Mumbai Indians|
|  419157|     2|   16|   5|      V Kohli|R Vinay Kumar|CRD Fernando|           1|         0|         1|           0|        0|         

In [ ]:
# Reading the matches DataFrame
df_matches = spark.read.csv('/content/drive/MyDrive/Colab Notebooks/DataSet/ipl_matches.csv', header=True, inferSchema=True)

# printing the Schema
df_matches.printSchema()

root
 |-- match_id: integer (nullable = true)
 |-- date: string (nullable = true)
 |-- player_of_match: string (nullable = true)
 |-- venue_id: integer (nullable = true)
 |-- neutral_venue: integer (nullable = true)
 |-- team1: string (nullable = true)
 |-- team2: string (nullable = true)
 |-- toss_winner: string (nullable = true)
 |-- toss_decision: string (nullable = true)
 |-- winner: string (nullable = true)
 |-- result: string (nullable = true)
 |-- result_margin: string (nullable = true)
 |-- eliminator: string (nullable = true)
 |-- method: string (nullable = true)
 |-- umpire1: string (nullable = true)
 |-- umpire2: string (nullable = true)



In [ ]:
# Printing the top 5 records
df_matches.show(5)

+--------+----------+---------------+--------+-------------+--------------------+--------------------+--------------------+-------------+--------------------+-------+-------------+----------+------+---------+--------------+
|match_id|      date|player_of_match|venue_id|neutral_venue|               team1|               team2|         toss_winner|toss_decision|              winner| result|result_margin|eliminator|method|  umpire1|       umpire2|
+--------+----------+---------------+--------+-------------+--------------------+--------------------+--------------------+-------------+--------------------+-------+-------------+----------+------+---------+--------------+
|  335982|2008-04-18|    BB McCullum|      35|            0|Royal Challengers...|Kolkata Knight Ri...|Royal Challengers...|        field|Kolkata Knight Ri...|   runs|          140|         N|    NA|Asad Rauf|   RE Koertzen|
|  335983|2008-04-19|     MEK Hussey|      31|            0|     Kings XI Punjab| Chennai Super Kings| C

In [ ]:
# adding a venue dataframe into a cluster
df_venue.createOrReplaceTempView('venue_df')

In [ ]:
# adding a ipl dataframe into a cluster
df_ipl_ball.createOrReplaceTempView('ipl_df')

In [ ]:
# adding a match dataframe into a cluster
df_matches.createOrReplaceTempView('match_df')

In [ ]:
# printing the list of dataframe inside the catalog cluster
spark.catalog.listTables()

[Table(name='ipl_df', database=None, description=None, tableType='TEMPORARY', isTemporary=True),
 Table(name='match_df', database=None, description=None, tableType='TEMPORARY', isTemporary=True),
 Table(name='venue_df', database=None, description=None, tableType='TEMPORARY', isTemporary=True)]

In [ ]:
# importing plotly library for visualization
import plotly.express as px

# **SECTION 2**

## Question 1:
Fine the top 3 venues which hosted the most number of eliminators?

In [ ]:
# joining mathces and venue dataframes
new_df = df_venue.join(df_matches, on='venue_id', how='inner')

In [ ]:
# filtering where the eliminator is yes
# selecting venue and eliminator features only
# groupby count based on venue and add to to new dataframe
new_ans_df1 = new_df.filter(new_df['eliminator']=='Y').select(['venue', 'eliminator']).groupBy('venue').count()

# visualizing the five records from the dataframe
new_ans_df1.show(5)

+--------------------+-----+
|               venue|count|
+--------------------+-----+
|Dubai Internation...|    3|
|Sardar Patel Stad...|    1|
|    Feroz Shah Kotla|    1|
|Rajiv Gandhi Inte...|    1|
|M.Chinnaswamy Sta...|    1|
+--------------------+-----+
only showing top 5 rows



In [ ]:
panda1_plot = new_ans_df1.sort('count', ascending=False).toPandas()
fig = px.bar(panda1_plot, x='venue', y='count', title='Top most hosted venues')
fig.show()

In [ ]:
# Sorting based on count in the descending order and printing the results
new_ans_df1.sort('count', ascending=False).head(3)

[Row(venue='Dubai International Cricket Stadium', count=3),
 Row(venue='Sheikh Zayed Stadium', count=2),
 Row(venue='Rajiv Gandhi International Stadium, Uppal', count=1)]

### observations:
1. **Dubai International Cricket Stadium** is the first most hosted venue with frequency of **3**.
2. **Sheikh Zayes Stadium** is the second most hosted venue with frequency of **2**.
3. **Rajiv Gandhi International Stadium** is third most hosted venue with frequency of **1**.
4. After that all the venues are having same frequency of **1**.

## Question 2: 
Return most number of catches taken by a player in IpL history?

In [ ]:
# capturing the where dissmissal kind is caught and add it in new dataframe
df_caught = df_ipl_ball.filter(df_ipl_ball['dismissal_kind']=='caught')

In [ ]:
# groupby count on fielder and add it on another dataframe
caught_count = df_caught.select('fielder').groupBy('fielder').count()

In [ ]:
panda2_plot = caught_count.sort('count', ascending=False).toPandas()
fig = px.bar(panda2_plot, x='fielder', y='count', title='Most Number of catche taking players')
fig.show()

In [ ]:
# sorting based on count and visualizing the top first record.  That is the answer
caught_count.sort('count', ascending=False).head(1)

[Row(fielder='KD Karthik', count=118)]

### observations:
1. Most number of catches taken by a player in IPL history is **KD Karthik** and the number of wicket he has taken is **118**

## Question 3:
1. write a query to return a report for highest wicket taker in matches which were affected by Duckworth-lewis's method(D/L)

In [ ]:
# joining the matches and ipl_ball dataframes on match id using inner join
new_df = df_matches.join(df_ipl_ball, on='match_id', how='inner')

# visualizing the top 2 records
new_df.show(2)

+--------+----------+---------------+--------+-------------+--------------------+--------------+--------------------+-------------+--------------+------+-------------+----------+------+---------------+----------+------+-----+----+-------+-------------+------------+------------+----------+----------+------------+---------+--------------+----------------+-------+-----------+--------------------+--------------+
|match_id|      date|player_of_match|venue_id|neutral_venue|               team1|         team2|         toss_winner|toss_decision|        winner|result|result_margin|eliminator|method|        umpire1|   umpire2|inning|overs|ball|batsman|  non_striker|      bowler|batsman_runs|extra_runs|total_runs|non_boundary|is_wicket|dismissal_kind|player_dismissed|fielder|extras_type|        batting_team|  bowling_team|
+--------+----------+---------------+--------+-------------+--------------------+--------------+--------------------+-------------+--------------+------+-------------+---------

In [ ]:
# filtering where the D/L method is there and put it to another DataFrame
method_df = new_df.filter(new_df['method']=='D/L')

In [ ]:
# plotting
pandas3_plot = method_df.select(['bowler', 'fielder', 'dismissal_kind']).filter(method_df['is_wicket']==1).groupBy(
    ['bowler', 'fielder', 'dismissal_kind']).count().sort('count', ascending=False).toPandas()
fig = px.bar(pandas3_plot, x='bowler', y='count', title='Most Number of wicket taker')
fig.show()

In [ ]:
# groupby count on bowler, fielder, dissmissal_kind.  and sorting it into ascending order and printing the top5 records
method_df.select(['bowler', 'fielder', 'dismissal_kind']).filter(method_df['is_wicket']==1).groupBy(
    ['bowler', 'fielder', 'dismissal_kind']).count().sort('count', ascending=False).head(5)

[Row(bowler='M Ntini', fielder='NA', dismissal_kind='bowled', count=3),
 Row(bowler='YS Chahal', fielder='AB de Villiers', dismissal_kind='caught', count=2),
 Row(bowler='MC Henriques', fielder='DW Steyn', dismissal_kind='caught', count=2),
 Row(bowler='JD Unadkat', fielder='RA Tripathi', dismissal_kind='caught', count=2),
 Row(bowler='NM Coulter-Nile', fielder='SA Yadav', dismissal_kind='caught', count=2)]

In [ ]:
pandas3_plot3 = new_df.select(['bowler', 'fielder', 'dismissal_kind']).filter(new_df['is_wicket']==1).groupBy(
    ['bowler', 'fielder', 'dismissal_kind']).count().sort('count', ascending=False).toPandas()
fig = px.bar(pandas3_plot3, x='bowler', y='count', title='most number of wicket taking bowler')
fig.show()

In [ ]:
##  ANATHER APPROACH TO THIS
## here we take all the methods D\L and NA both method.
## becasue i don't know which one is right.
new_df.select(['bowler', 'fielder', 'dismissal_kind']).filter(new_df['is_wicket']==1).groupBy(
    ['bowler', 'fielder', 'dismissal_kind']).count().sort('count', ascending=False).head(5)

[Row(bowler='SL Malinga', fielder='NA', dismissal_kind='bowled', count=63),
 Row(bowler='PP Chawla', fielder='NA', dismissal_kind='bowled', count=43),
 Row(bowler='SP Narine', fielder='NA', dismissal_kind='bowled', count=36),
 Row(bowler='B Kumar', fielder='NA', dismissal_kind='bowled', count=36),
 Row(bowler='RA Jadeja', fielder='NA', dismissal_kind='bowled', count=29)]

### observations:
1. Highest number of wicket taker is **SL Malinga**. wicket type is **bowled**. wicket count is **63**. IT BASED ON SECOND RESULT
2. Highest number of wicket taker is **M Ntini**. wicket type is **bowled**. wicket count is **3**.  IT BASED ON FIRST RESULT

## Question 4 
Write a query to return a report for highest strike rate by a batsman in non powerplay
overs(7-20 overs)
Note: strike rate = (Total Runs scored/Total balls faced by player) *100, Make sure that balls
faced by players should be legal delivery (not wide balls or no balls).

In [ ]:
# creating a dataframe by filtering normal delivery balls and non powerplays 
normal_ball_df = df_ipl_ball.filter((df_ipl_ball['extras_type'] == 'NA') & (df_ipl_ball['overs']>7))

In [ ]:
# using groupby sum on batsman and creating a new Dataframe
sum_ipl_df = normal_ball_df.groupBy('batsman').sum()

In [ ]:
# printing the summed DataFrame
sum_ipl_df.show()

+--------------+-------------+-----------+----------+---------+-----------------+---------------+---------------+-----------------+--------------+
|       batsman|sum(match_id)|sum(inning)|sum(overs)|sum(ball)|sum(batsman_runs)|sum(extra_runs)|sum(total_runs)|sum(non_boundary)|sum(is_wicket)|
+--------------+-------------+-----------+----------+---------+-----------------+---------------+---------------+-----------------+--------------+
| Kuldeep Yadav|     84918193|        114|      1229|      272|               57|              0|             57|                0|             4|
|    TM Dilshan|    192904214|        638|      5270|     1598|              540|              0|            540|                0|            17|
|    S Anirudha|     37642895|         89|      1147|      301|              100|              0|            100|                0|             5|
|    KA Pollard|   1524760433|       2583|     26799|     6914|             2904|              0|           2904|     

In [ ]:
# selecting the each batsman and finding the strike rate the and creating a dataframe accordingly
ans4_df = sum_ipl_df.withColumn('STRIKE_RATE', (sum_ipl_df['sum(total_runs)']/sum_ipl_df['sum(ball)'])*100).select(['batsman', 'STRIKE_RATE'])

# visualizing the results
ans4_df.sort('STRIKE_RATE', ascending=False).show(3)

+------------+------------------+
|     batsman|       STRIKE_RATE|
+------------+------------------+
|   DJ Jacobs|133.33333333333331|
|Vishnu Vinod|              75.0|
|  ER Dwivedi| 70.37037037037037|
+------------+------------------+
only showing top 3 rows



In [ ]:
# plotting the strike rate of batsman using plotly
fig = px.bar(ans4_df.sort('STRIKE_RATE', ascending=False).toPandas(), x='batsman', y='STRIKE_RATE', title='Strike Rates of a batsman')
fig.show()

### observations:
1.  Batsman who is having the highest strike rate is **DJ Jacobs**.
2.  Strike Rate is **133.33**.

## Question 5
Write a query to return a report for highest extra runs in a venue(Stadium, City)

In [ ]:
## joining the new and venue both the  dataframes and creating new dataframe
new_df1 = new_df.join(df_venue, on='venue_id', how='inner')

In [ ]:
# showing the top 5 records of the new dataframe
new_df1.show(5)

+--------+--------+----------+---------------+-------------+--------------------+--------------+--------------------+-------------+--------------+------+-------------+----------+------+---------------+----------+------+-----+----+-------------+-------------+------------+------------+----------+----------+------------+---------+--------------+----------------+-------+-----------+--------------------+--------------+--------------------+---------+
|venue_id|match_id|      date|player_of_match|neutral_venue|               team1|         team2|         toss_winner|toss_decision|        winner|result|result_margin|eliminator|method|        umpire1|   umpire2|inning|overs|ball|      batsman|  non_striker|      bowler|batsman_runs|extra_runs|total_runs|non_boundary|is_wicket|dismissal_kind|player_dismissed|fielder|extras_type|        batting_team|  bowling_team|               venue|     city|
+--------+--------+----------+---------------+-------------+--------------------+--------------+------

In [ ]:
# selecting the venue, city, extran_runs columns and groupby sum based on venue, city storing it into another dataframe
ans5_df = new_df1.select(['venue', 'city', 'extra_runs']).groupBy(['venue', 'city']).sum()

# sorting the Data in the asceind order and visualizing the top 3 records
ans5_df.sort('sum(extra_runs)', ascending=False).head(3)

[Row(venue='M.Chinnaswamy Stadium', city='Bengaluru', sum(extra_runs)=1355),
 Row(venue='Wankhede Stadium', city='Mumbai', sum(extra_runs)=1232),
 Row(venue='Eden Gardens', city='Kolkata', sum(extra_runs)=1215)]

In [ ]:
# plotting the venues where highes extra runs are present
fig = px.bar(ans5_df.sort('sum(extra_runs)', ascending=False).toPandas(), x='venue', y='sum(extra_runs)', title="venue where highest extra run present")
fig.show()

### observations:
1. Highest extra run in **M.chinnaswamy Stadium**, city=**Bengaluru**, extra_runs=**1355**.

## Question 6: 
write a query to return a report for the cricketers with the most number of players of the match award in the neutral venues.

In [ ]:
# selecting where the neutral_venue == 1 and put in into a separate DataFrame
neutral_df = df_matches.filter(df_matches['neutral_venue']==1)

In [ ]:
# groupby the players and based on the count sorting it printing the tail 20.
players = neutral_df.groupby('player_of_match').count()
players.sort('count', ascending=False).head(5)

[Row(player_of_match='GJ Maxwell', count=3),
 Row(player_of_match='JH Kallis', count=3),
 Row(player_of_match='YK Pathan', count=3),
 Row(player_of_match='JP Duminy', count=3),
 Row(player_of_match='DR Smith', count=2)]

In [ ]:
# plotting the highest player of the award winning players
pandas6_plot = players.sort('count', ascending=False).toPandas()
fig = px.bar(pandas6_plot, x='player_of_match', y='count', title='Highes player of the award')
fig.show()

### observations:
name of cricketers who got highest player award in neutral venue:
1. **YK Pathan**  -> number of awards **3**
2. **GJ Maxwell**    -> number of awards **3**
3. **JH Kallis**  -> number of awards **3**
4. **JP Duminy**  -> number of awards **3**


## Question 7:
Write a query to get a list of top 10 players with the highest batting average.
NOTE: Batting average is the total number of runs scored divided by the number of times they have been out (Make sure to include runout (on Non-Striker end) as valid out while calculating average).


In [ ]:
# showing the ipl ball by bll dataframe
df_ipl_ball.show(5)

+--------+------+-----+----+-------------+-------------+------------+------------+----------+----------+------------+---------+--------------+----------------+-------+-----------+--------------------+--------------+
|match_id|inning|overs|ball|      batsman|  non_striker|      bowler|batsman_runs|extra_runs|total_runs|non_boundary|is_wicket|dismissal_kind|player_dismissed|fielder|extras_type|        batting_team|  bowling_team|
+--------+------+-----+----+-------------+-------------+------------+------------+----------+----------+------------+---------+--------------+----------------+-------+-----------+--------------------+--------------+
|  419157|     2|   16|   4|      V Kohli|R Vinay Kumar|CRD Fernando|           0|         0|         0|           0|        0|            NA|              NA|     NA|         NA|Royal Challengers...|Mumbai Indians|
|  419157|     2|   16|   5|      V Kohli|R Vinay Kumar|CRD Fernando|           1|         0|         1|           0|        0|         

In [ ]:
# Take a total number of scores scored by a batsman and put it in a separate dataframe named scores
scores = df_ipl_ball.select(['batsman', 'total_runs']).groupBy('batsman').sum().sort('sum(total_runs)')
scores

DataFrame[batsman: string, sum(total_runs): bigint]

In [ ]:
# Take a total number of time a batsman out as a batsman and put it in striker dataframe
striker = df_ipl_ball.filter(((df_ipl_ball['dismissal_kind']!='NA')&(df_ipl_ball['dismissal_kind']!='retired hurt'))&(
    df_ipl_ball.player_dismissed==df_ipl_ball.batsman)).select(
    ['batsman', 'dismissal_kind']).groupBy('batsman').count()
striker.show(5)

+-------------+-----+
|      batsman|count|
+-------------+-----+
|   TM Dilshan|   40|
|   S Anirudha|    8|
|Kuldeep Yadav|    4|
|   KA Pollard|  100|
|      J Botha|   20|
+-------------+-----+
only showing top 5 rows



In [ ]:
# Take a total number of times a batsman out as a non striker and put it in separate dataframe named non_striker
non_striker = df_ipl_ball.filter((df_ipl_ball.dismissal_kind=='run out') & (df_ipl_ball.player_dismissed==df_ipl_ball.non_striker
                                                              )).groupBy('non_striker').count()
non_striker.show(5)

+-------------+-----+
|  non_striker|count|
+-------------+-----+
|   TM Dilshan|    3|
|Kuldeep Yadav|    2|
|   KA Pollard|    1|
|LA Carseldine|    1|
|     DR Smith|    4|
+-------------+-----+
only showing top 5 rows



In [ ]:
# Renaming the columns for merging
non_striker = non_striker.withColumnRenamed('non_striker', 'batsman')
non_striker = non_striker.withColumnRenamed('count', 'count1')

In [ ]:
# joining the striker and non striker DataFrames
new_df7 = striker.join(non_striker, on='batsman', how='inner')
new_df7.show(5)

+-------------+-----+------+
|      batsman|count|count1|
+-------------+-----+------+
|   TM Dilshan|   40|     3|
|Kuldeep Yadav|    4|     2|
|   KA Pollard|  100|     1|
|LA Carseldine|    3|     1|
|     DR Smith|   80|     4|
+-------------+-----+------+
only showing top 5 rows



In [ ]:
### calculating the total number of outs and put it in a new columns
new_df7 = new_df7.withColumn('total_out', new_df7['count']+new_df7['count1'])
new_df7.show(5)

+-------------+-----+------+---------+
|      batsman|count|count1|total_out|
+-------------+-----+------+---------+
|   TM Dilshan|   40|     3|       43|
|Kuldeep Yadav|    4|     2|        6|
|   KA Pollard|  100|     1|      101|
|LA Carseldine|    3|     1|        4|
|     DR Smith|   80|     4|       84|
+-------------+-----+------+---------+
only showing top 5 rows



In [ ]:
# mering the scores and new_df7 dataframe for calculating batting average
ans_df7 = new_df7.join(scores, on='batsman', how='inner')
ans_df7.show(5)

+-------------+-----+------+---------+---------------+
|      batsman|count|count1|total_out|sum(total_runs)|
+-------------+-----+------+---------+---------------+
|   TM Dilshan|   40|     3|       43|           1251|
|Kuldeep Yadav|    4|     2|        6|             59|
|   KA Pollard|  100|     1|      101|           3188|
|LA Carseldine|    3|     1|        4|             87|
|     DR Smith|   80|     4|       84|           2515|
+-------------+-----+------+---------+---------------+
only showing top 5 rows



In [ ]:
ans_df7 = ans_df7.withColumn('Batting_avg', ans_df7['sum(total_runs)']/ans_df7['total_out'])

### drop the unwanted features for clear visualization
ans_df7 = ans_df7.drop('count')
ans_df7 = ans_df7.drop('count1')

ans_df7.sort('Batting_avg', ascending=False).show(10)

+--------------+---------+---------------+------------------+
|       batsman|total_out|sum(total_runs)|       Batting_avg|
+--------------+---------+---------------+------------------+
|    RD Gaikwad|        4|            220|              55.0|
| Iqbal Abdulla|        2|             97|              48.5|
|     DA Warner|      123|           5522| 44.89430894308943|
|      CH Gayle|      116|           5103|43.991379310344826|
|      MS Dhoni|      113|           4855|  42.9646017699115|
|PD Collingwood|        5|            213|              42.6|
|      SE Marsh|       62|           2598|41.903225806451616|
|AB de Villiers|      120|           5016|              41.8|
|     JP Duminy|       51|           2131| 41.78431372549019|
|   LMP Simmons|       27|           1127| 41.74074074074074|
+--------------+---------+---------------+------------------+
only showing top 10 rows



In [ ]:
# plotting the Batting average of batsmans
pandas7_plot = ans_df7.sort('Batting_avg', ascending=False).toPandas()
fig = px.bar(pandas7_plot, x='batsman', y='Batting_avg', title='Batting average of batsmans')
fig.show()

### observations:
above are top ten batsmans having the highest batting average.
1.  **RD Gaikwad** Batting average is **55.0**
2.  **Iqbal Abdulla** Batting average is **48.5**
3.  **DA Warner**  Batting average is **44.89**
4.  **CH Gayle** Batting average is **43.99**
5.  **MS Dhoni** Batting average is **42.96**
6.  **PD Collingwood** Batting average is **42.6**
7.  **SE Marsh** Batting average is **41.90**
8.  **AB de villiers** Batting average is **41.8**
9.  **JP Duminy** Batting average is **41.78**
10. **LMP Simmons** Batting average is **41.74**



## Question 8:
write a query to find out who has officiated as an(umpire) the most number of matches in IPL.

In [ ]:
# taking the frequency of umpire 1 and umpire 2
count_umpire1 = df_matches.select('umpire1').groupBy('umpire1').count()
count_umpire2 = df_matches.select('umpire2').groupBy('umpire2').count()
print(count_umpire1.show(3))
print()
print(count_umpire2.show(3))

+--------------+-----+
|       umpire1|count|
+--------------+-----+
|   RE Koertzen|   20|
|      SJ Davis|    6|
|AV Jayaprakash|    2|
+--------------+-----+
only showing top 3 rows

None

+--------------+-----+
|       umpire2|count|
+--------------+-----+
|   RE Koertzen|   21|
|      SJ Davis|    1|
|AV Jayaprakash|    4|
+--------------+-----+
only showing top 3 rows

None


In [ ]:
# Renaming the column in count_umpire2 dataframe for merging
count_umpire2 = count_umpire2.withColumnRenamed('count', 'count1')
count_umpire2 = count_umpire2.withColumnRenamed('umpire2', 'umpire1')
count_umpire2.show(3)

+--------------+------+
|       umpire1|count1|
+--------------+------+
|   RE Koertzen|    21|
|      SJ Davis|     1|
|AV Jayaprakash|     4|
+--------------+------+
only showing top 3 rows



In [ ]:
# merging the count_umpire1 and count_umpire2 Dataframe
ans_df8 = count_umpire1.join(count_umpire2, on='umpire1', how='inner')
ans_df8.show(3)

+--------------+-----+------+
|       umpire1|count|count1|
+--------------+-----+------+
|   RE Koertzen|   20|    21|
|      SJ Davis|    6|     1|
|AV Jayaprakash|    2|     4|
+--------------+-----+------+
only showing top 3 rows



In [ ]:
# creating the new column total count and putting the sum of count1 and count 2 over there
ans_df8 = ans_df8.withColumn('total_count', ans_df8['count']+ans_df8['count1'])
ans_df8.show(3)

+--------------+-----+------+-----------+
|       umpire1|count|count1|total_count|
+--------------+-----+------+-----------+
|   RE Koertzen|   20|    21|         41|
|      SJ Davis|    6|     1|          7|
|AV Jayaprakash|    2|     4|          6|
+--------------+-----+------+-----------+
only showing top 3 rows



In [ ]:
## showing the results
ans_df8.sort('total_count', ascending=False).show(5)

+---------------+-----+------+-----------+
|        umpire1|count|count1|total_count|
+---------------+-----+------+-----------+
|         S Ravi|   37|    84|        121|
|HDPK Dharmasena|   78|    16|         94|
|   AK Chaudhary|   56|    31|         87|
|  C Shamshuddin|   22|    60|         82|
|      M Erasmus|   40|    25|         65|
+---------------+-----+------+-----------+
only showing top 5 rows



In [ ]:
fig = px.bar(ans_df8.sort('total_count', ascending=False).toPandas(), x='umpire1', y='total_count', title='most umpire count')
fig.show()

### observations:
1. **S Ravi** has officiated as an umpire most number of matches. match count is **121**

# Quesiont 9 :
Find venue Details of the match where V Kohli Scored his highest individual runs in ipl.


In [ ]:
from pyspark.sql import functions as func

In [ ]:
match_id = df_ipl_ball.filter(df_ipl_ball.batsman == 'V Kohli').groupBy('match_id').agg(
    func.col('match_id'), func.sum('batsman_runs')).sort('sum(batsman_runs)', ascending=False).collect()[0]['match_id']
match_id

980999

In [ ]:
df_venue.filter(df_venue['venue_id']==df_matches.filter(df_matches.match_id == match_id).collect()[0]['venue_id']).head()

Row(venue_id=35, venue='M.Chinnaswamy Stadium', city='Bengaluru')

### observations:
1. V kohli Scored his highest ipl score **M.Chinnaswamy Stadium**. city is **Bengaluru**. venue_id is **35** 

## Question 10: Creating Case Study
Please Analys how winning/lossing tosses can impact a match and it's result?


### in pyspark dataframe we don't have a visualization plots.
so we can convert pyspark dataframe to pandas for visualization only.

In [ ]:
df_matches.show(5)

+--------+----------+---------------+--------+-------------+--------------------+--------------------+--------------------+-------------+--------------------+-------+-------------+----------+------+---------+--------------+
|match_id|      date|player_of_match|venue_id|neutral_venue|               team1|               team2|         toss_winner|toss_decision|              winner| result|result_margin|eliminator|method|  umpire1|       umpire2|
+--------+----------+---------------+--------+-------------+--------------------+--------------------+--------------------+-------------+--------------------+-------+-------------+----------+------+---------+--------------+
|  335982|2008-04-18|    BB McCullum|      35|            0|Royal Challengers...|Kolkata Knight Ri...|Royal Challengers...|        field|Kolkata Knight Ri...|   runs|          140|         N|    NA|Asad Rauf|   RE Koertzen|
|  335983|2008-04-19|     MEK Hussey|      31|            0|     Kings XI Punjab| Chennai Super Kings| C

In [ ]:
# filtering where the toss winner is match winner
toss_mat_win = df_matches.select(['toss_winner', 'toss_decision', 'winner', 'result_margin']).filter(df_matches.toss_winner==df_matches.winner)

# filtering where the toss winner is not match winner
toss_mat_loss = df_matches.select(['toss_winner', 'toss_decision', 'winner', 'result_margin']).filter(df_matches.toss_winner!=df_matches.winner)

In [ ]:
# adding a new column with value for the analysis
toss_mat_win = toss_mat_win.withColumn('win1', toss_mat_win.result_margin-toss_mat_win.result_margin+1)

# adding a new column for the analysis
toss_mat_loss = toss_mat_loss.withColumn('win1', toss_mat_loss.result_margin-toss_mat_loss.result_margin+1)

In [ ]:
# filtering the toss_decision==bat from the  toss winner == match_winner
toss_win = toss_mat_win.filter(toss_mat_win.toss_decision=='bat')

# filtering the toss_decision == bat from the toss winner == match_loser
toss_loss = toss_mat_loss.filter(toss_mat_loss.toss_decision=='bat')

In [ ]:
# grouping with respect to toss inners
toss_win_group = toss_win.groupBy('toss_winner').sum()
toss_loss_group = toss_loss.groupBy('toss_winner').sum()

In [ ]:
# converting it into pandas DataFrame visualization
toss_win1 = toss_win_group.toPandas()
toss_loss1 = toss_loss_group.toPandas()

In [ ]:
toss_win1.head(1)

toss_winner  sum(win1)
0  Sunrisers Hyderabad       10.0

In [ ]:
toss_loss1.head(1)

toss_winner  sum(win1)
0  Sunrisers Hyderabad       14.0

In [ ]:
# total number of winning count for toss winners
win_count = toss_mat_win.groupBy('toss_winner').sum()

# total number of losing count for toss winners
loss_count = toss_mat_win.groupBy('toss_winner').sum()

In [ ]:
win_count1 = win_count.toPandas()
loss_count1 = loss_count.toPandas()

In [ ]:
win_count1.head(1)

toss_winner  sum(win1)
0  Sunrisers Hyderabad       29.0

In [ ]:
loss_count1.head(1)

toss_winner  sum(win1)
0  Sunrisers Hyderabad       29.0

In [ ]:
# Bar chart for where the toss winner == match_winner
import plotly.express as px
fig = px.bar(win_count1, x='toss_winner', y='sum(win1)', title='toss_winner is  match_winner')
fig.show()

In [ ]:
## # Bar chart for where the toss winner != match_winner
import plotly.express as px
fig = px.bar(loss_count1, x='toss_winner', y='sum(win1)', title='toss_winner is not a match_winner')
fig.show()

### observations:
1. above 2 plots we can conclude that.  Toss winners are winning the match and Toss winners are losing the match both are equal.  So there is no impact with respect to toss.  We can start analyzing the toss decision.

### Here we are checking whether there is any change in result with respect to toss_decision.

In [ ]:
import plotly.express as px
fig = px.bar(toss_win1, x='toss_winner', y='sum(win1)', title='toss_winner==match_winner and toss_decition == bat')
fig.show()

In [ ]:
import plotly.express as px
fig = px.bar(toss_loss1, x='toss_winner', y='sum(win1)', title='toss_winner is not match_loser and toss_decition == bat')
fig.show()

### observations:
1.  when compared to 1 st plot we can clearly see that some of the teams losing because of choosing the batting.  There is impact with respect to toss_decision.  Further we can start analysing it.

## Question 10:
part 2 analysis.  in the below analysis we are going to predict the percentage of winning with respect to toss decision for each team. 

In [ ]:
# selecting where the toss winning team is match winning team also toss decision is bat
batting = df_matches.select(['toss_winner', 'toss_decision', 'winner']).filter((df_matches['toss_winner']==df_matches['winner']) & 
                                                                     (df_matches['toss_decision']=='bat')).select('toss_winner').groupBy('toss_winner').count().toPandas()
# selecting where the toss winning team is match winning team also toss decision is bat                                                             
fielding = df_matches.select(['toss_winner', 'toss_decision', 'winner']).filter((df_matches['toss_winner']==df_matches['winner']) & 
                                                                     (df_matches['toss_decision']=='field')).select('toss_winner').groupBy('toss_winner').count().toPandas()


In [ ]:
total_count = df_matches.select('toss_winner').groupBy('toss_winner').count().toPandas()

In [ ]:
batting['percentage_winning'] = batting['count']/total_count['count']
fielding['percentage_winning'] = fielding['count']/total_count['count']

In [ ]:
fig_final = px.bar(batting, x='toss_winner', y='percentage_winning', title = 'Percentage of team win where the toss decision is bat')
fig_final.show()

In [ ]:
fig_final1 = px.bar(fielding, x='toss_winner', y='percentage_winning', title = 'Percentage of team win where the toss decision is fielding')
fig_final1.show()

### observations:
1. from the above 2 plots we can clearly see that.  team who won the toss and choosing the batting less number of winning over there.  teams who won the toss and choosing the fielding having the more winning.  But the percentage of winning is high who won the toss and winning the match.  from the toss winning we can assume that teams who didn't won the toss and and given fielding has the huge number of winning and less number of winning percentage.  and teams who did't won the toss and give bat has the high numberr of winning and less number of winning percentage.
2. Finally we can conclude that toss winning is not impact the result of the match.  who won the toss or doesn't won the toss propbablity of winning and losing is equal.  But when compared to toss decision.  it impacts the result of the match.  some of the teams like deccan challengers royal challengers Bengaluru having the highest probablity of winning because of toss decision is batting.  but some teams are losing the match.  because of batting.  But whover take the batting has the highest probablity of winning and less number of losing.

# SECTION 3

In [ ]:
from pyspark.sql import SparkSession
import plotly.express as px

# Creating a class database
class Database:

  # Creating a constructor for a class
  # constructor take a value from a csv file
  def __init__(self, appName='Work'):
    self.spark = (SparkSession
    .builder
    .appName(appName)
    .getOrCreate())

    # loading the vanue dataset
    try:
      self.df_venue = self.spark.read.csv('/content/drive/MyDrive/Colab Notebooks/DataSet/ipl_venue.csv', header=True, inferSchema=True)
    except:
      print("Can't load venue dataset. pleach check file path and syntax")
    
    # loading the ipl_ball_by_ball dataset
    try:
      self.df_ipl_ball = self.spark.read.csv('/content/drive/MyDrive/Colab Notebooks/DataSet/ipl_ball_by_ball.csv', header=True, inferSchema=True)
    except:
      print("Can't load ipl_ball by ball dataset. please check file path and syntax")
    
    # loading the matches dataset
    try:
      self.df_matches = self.spark.read.csv('/content/drive/MyDrive/Colab Notebooks/DataSet/ipl_matches.csv', header=True, inferSchema=True)
    except:
      print("can't load matches dataset. please check the file path and syntax")

    # putting all the data inside the cluster using catalog
    self.df_venue.createOrReplaceTempView('venue_df')
    self.df_ipl_ball.createOrReplaceTempView('ipl_df')
    self.df_matches.createOrReplaceTempView('match_df')
    #self.spark.catalog.listTables()


  # return the answer to Q1 in Section 2 - return the answer as a dictionary
  def get_query1_result(self):
    try:
      self.new_df = self.df_venue.join(self.df_matches, on='venue_id', how='inner')
      self.new_ans_df1 = self.new_df.filter(self.new_df['eliminator']=='Y').select(['venue', 'eliminator']).groupBy('venue').count()
      self.ans = self.new_ans_df1.sort('count', ascending=False).select("venue").toPandas()
      self.panda1_plot = self.new_ans_df1.sort('count', ascending=False).toPandas()
      self.fig = px.bar(panda1_plot, x='venue', y='count', title='Top most hosted venues')
      self.fig.show()
      return {"Q1" : list(self.ans.venue.values)[:3]}
    except:
      print('Something is Wrong with Query1.  Please check the Query1 code, and try to execute section2.')

  # return the answer to Q2 in Section 2 - return the answer as a dictionary
  def get_query2_result(self):
    try:
      self.df_caught = self.df_ipl_ball.where(self.df_ipl_ball['dismissal_kind'] == 'caught').select('fielder').groupBy('fielder').count()
      self.query2 = self.df_caught.sort('count', ascending=False).select('fielder').toPandas()
      self.panda2_plot = self.query2
      self.fig = px.bar(panda2_plot, x='fielder', y='count', title='Most Number of catche taking players')
      self.fig.show()
      return {'Q2':list(self.query2.fielder.values)[0]}
    except:
      print('Somethig is wrong with the Query2 code. please check the code and syntax and try to execute section2')

  # return the answer to Q3 in Section 2 - return the answer as a dictionary
  def get_query3_result(self):
    try:
      self.new_df = self.df_matches.join(df_ipl_ball, on='match_id', how='inner')
      self.method_df = self.new_df.filter(self.new_df['method']=='D/L')

      self.pandas3_plot = self.method_df.select(['bowler', 'fielder', 'dismissal_kind']).filter(self.method_df['is_wicket']==1).groupBy(
      ['bowler', 'fielder', 'dismissal_kind']).count().sort('count', ascending=False).toPandas()
      self.fig = px.bar(pandas3_plot, x='bowler', y='count', title='Most Number of wicket taker')
      self.fig.show()
      self.query3 = self.method_df.select(['bowler', 'fielder', 'dismissal_kind']).filter(self.method_df['is_wicket']==1).groupBy(
      ['bowler', 'fielder', 'dismissal_kind']).count().sort('count', ascending=False).toPandas()

      self.pandas3_plot3 = self.new_df.select(['bowler', 'fielder', 'dismissal_kind']).filter(new_df['is_wicket']==1).groupBy(
      ['bowler', 'fielder', 'dismissal_kind']).count().sort('count', ascending=False).toPandas()
      self.fig = px.bar(pandas3_plot3, x='bowler', y='count', title='most number of wicket taking bowler')
      self.fig.show()

      self.ans33 = self.new_df.select(['bowler', 'fielder', 'dismissal_kind']).filter(new_df['is_wicket']==1).groupBy(
      ['bowler', 'fielder', 'dismissal_kind']).count().sort('count', ascending=False).toPandas()

      return {'Q3 most wicket taker ':list(self.query3.bowler.values)[0], "Q3 most wicket taking bowler:":list(self.ans33.bowler.values)[0]}
    except:
      print('Something is wrong with the query3 code.  Please Check and correct it. and try to execute section2')

  # return the answer to Q4 in Section 2 - return the answer as a dictionary
  def get_query4_result(self):
    try:
      self.normal_ball_df = self.df_ipl_ball.filter((self.df_ipl_ball['extras_type'] == 'NA') & (self.df_ipl_ball['overs']>7))
      self.sum_ipl_df = self.normal_ball_df.groupBy('batsman').sum()
      self.ans4_df = self.sum_ipl_df.withColumn('STRIKE_RATE', (self.sum_ipl_df['sum(total_runs)']/self.sum_ipl_df['sum(ball)'])*100).select(['batsman', 'STRIKE_RATE'])
      self.query4 = self.ans4_df.sort('STRIKE_RATE', ascending=False).select(['batsman', 'STRIKE_RATE']).toPandas()
      self.fig = px.bar(self.ans4_df.sort('STRIKE_RATE', ascending=False).toPandas(), x='batsman', y='STRIKE_RATE', title='Strike Rates of a Batsman')
      self.fig.show()
      return {'Q4':list(self.query4.batsman.values)[0], 'Q4i':list(self.query4.STRIKE_RATE.values)[0]}
    except:
      print('Something is wrong with the query4 code.  Please check and correct it and try to execute section 2')

  # return the answer to Q5 in Section 2 - return the answer as a dictionary
  def get_query5_result(self):
    try:
      self.new_df = self.df_matches.join(df_ipl_ball, on='match_id', how='inner')
      self.new_df1 = self.new_df.join(df_venue, on='venue_id', how='inner')
      self.ans5_df = self.new_df1.select(['venue', 'city', 'extra_runs']).groupBy(['venue', 'city']).sum()
      self.query5 = self.ans5_df.sort('sum(extra_runs)', ascending=False).select(['venue','city']).toPandas()
      self.fig = px.bar(self.ans5_df.sort('sum(extra_runs)', ascending=False).toPandas(), x='venue', y='sum(extra_runs)', title="venue where highest extra run present")
      self.fig.show()
      return {'Q5':list(self.query5.venue.values)[0], 'Q5i':list(self.query5.city.values)[0]}
    except:
      print('Something is wrong with the query5 code.  Please check and correct it and try to execute section2')

  # return the answer to Q6 in Section 2 - return the answer as a dictionary
  def get_query6_result(self):
    try:
      self.neutral_df = self.df_matches.filter(self.df_matches['neutral_venue']==1)
      self.players = self.neutral_df.groupby('player_of_match').count()
      self.query6 = self.players.sort('count', ascending=False).select('player_of_match').toPandas()
      self.pandas6_plot = self.players.sort('count', ascending=False).toPandas()
      self.fig = px.bar(pandas6_plot, x='player_of_match', y='count', title='Highest player of match award')
      self.fig.show()
      return {'Q6':list(self.query6.player_of_match.values)[0]}
    except:
      print('Something is wrong with the query6 code.  Please check and correct it and try to execute section2')

  # return the answer to Q7 in Section 2 - return the answer as a dictionary
  def get_query7_result(self):
    try:
      self.scores = self.df_ipl_ball.select(['batsman', 'total_runs']).groupBy('batsman').sum().sort('sum(total_runs)')
      self.striker = self.df_ipl_ball.filter(((self.df_ipl_ball['dismissal_kind']!='NA')&(self.df_ipl_ball['dismissal_kind']!='retired hurt'))&(
      self.df_ipl_ball.player_dismissed==self.df_ipl_ball.batsman)).select(
      ['batsman', 'dismissal_kind']).groupBy('batsman').count()
      self.non_striker = self.df_ipl_ball.filter((self.df_ipl_ball.dismissal_kind=='run out') & (self.df_ipl_ball.player_dismissed==self.df_ipl_ball.non_striker
                                                                )).groupBy('non_striker').count()
      self.non_striker = self.non_striker.withColumnRenamed('non_striker', 'batsman')
      self.non_striker = self.non_striker.withColumnRenamed('count', 'count1')
      self.new_df7 = self.striker.join(non_striker, on='batsman', how='inner')
      self.new_df7 = self.new_df7.withColumn('total_out', self.new_df7['count']+self.new_df7['count1'])
      self.ans_df7 = self.new_df7.join(scores, on='batsman', how='inner')
      self.ans_df7 = self.ans_df7.withColumn('Batting_avg', self.ans_df7['sum(total_runs)']/self.ans_df7['total_out'])
      self.ans_df7 = self.ans_df7.drop('count')
      self.ans_df7 = self.ans_df7.drop('count1')
      self.query7 = self.ans_df7.sort('Batting_avg', ascending=False).select('batsman').toPandas()
      self.pandas7_plot = ans_df7.sort('Batting_avg', ascending=False).toPandas()
      self.fig = px.bar(pandas7_plot, x='batsman', y='Batting_avg', title='Batting Avg of Batsmans')
      self.fig.show()
      return {"Q7": list(self.query7.batsman.values)[:10]}
    except:
      print('Something is wrong with the query7 code.  Please check and correct it. and try to execute section2.')

  # return the answer to Q8 in Section 2 - return the answer as a dictionary
  def get_query8_result(self):
    try:
      self.count_umpire1 = self.df_matches.select('umpire1').groupBy('umpire1').count()
      self.count_umpire2 = self.df_matches.select('umpire2').groupBy('umpire2').count()
      self.count_umpire2 = self.count_umpire2.withColumnRenamed('count', 'count1')
      self.count_umpire2 = self.count_umpire2.withColumnRenamed('umpire2', 'umpire1')
      self.ans_df8 = self.count_umpire1.join(count_umpire2, on='umpire1', how='inner')
      self.ans_df8 = self.ans_df8.withColumn('total_count', self.ans_df8['count']+self.ans_df8['count1'])
      self.query8 = self.ans_df8.sort('total_count', ascending=False).select('umpire1').toPandas()
      self.fig = px.bar(self.ans_df8.sort('total_count', ascending=False).toPandas(), x='umpire1', y='total_count')
      self.fig.show()
      return {'Q8': list(self.query8.umpire1.values)[0]}
    except:
      print('Something is wrong with the query8 code.  Please check and correct it and try to execute section 2')

  # return the answer to Q9 in Section 2 - return the answer as a dictionary
  def get_query9_result(self):
    try:
      self.match_id = self.df_ipl_ball.filter(self.df_ipl_ball.batsman == 'V Kohli').groupBy('match_id').agg(
        func.col('match_id'), func.sum('batsman_runs')).sort('sum(batsman_runs)', ascending=False).collect()[0]['match_id']
      self.query9 = self.df_venue.filter(self.df_venue['venue_id']==self.df_matches.filter(self.df_matches.match_id == match_id).collect()[0]['venue_id']).select(['venue', 'city']).toPandas()
      return {'Q9':list(self.query9.venue.values)[0], 'Q9i':list(self.query9.city.values)[0]}
    except:
      print('Something is wrong with the query9 code.  Please check and correct it and try to execute section 2')

  # return the aswer to Q10 in Section 2 - return the answer as a dictionary
  def get_query10_result(self):
    try:
      self.toss_mat_win = self.df_matches.select(['toss_winner', 'toss_decision', 'winner', 'result_margin']).filter(self.df_matches.toss_winner==self.df_matches.winner)
      self.toss_mat_loss = self.df_matches.select(['toss_winner', 'toss_decision', 'winner', 'result_margin']).filter(self.df_matches.toss_winner!=self.df_matches.winner)
      self.toss_mat_win = self.toss_mat_win.withColumn('win1', self.toss_mat_win.result_margin-self.toss_mat_win.result_margin+1)
      self.toss_mat_loss = self.toss_mat_loss.withColumn('win1', self.toss_mat_loss.result_margin-self.toss_mat_loss.result_margin+1)
      self.toss_win = self.toss_mat_win.filter(self.toss_mat_win.toss_decision=='bat')
      self.toss_loss = self.toss_mat_loss.filter(self.toss_mat_loss.toss_decision=='bat')
      self.toss_win_group = self.toss_win.groupBy('toss_winner').sum()
      self.toss_loss_group = self.toss_loss.groupBy('toss_winner').sum()
      self.toss_win1 = self.toss_win_group.toPandas()
      self.toss_loss1 = self.toss_loss_group.toPandas()
      self.fig = px.bar(win_count1, x='toss_winner', y='sum(win1)', title='toss_winner is  match_winner')
      self.fig.show()
      self.fig1 = px.bar(loss_count1, x='toss_winner', y='sum(win1)', title='toss_winner is not a match_winner')
      self.fig1.show()
      self.fig2 = px.bar(toss_win1, x='toss_winner', y='sum(win1)', title='toss_winner==match_winner and toss_decition == bat')
      self.fig2.show()
      self.fig3 = px.bar(toss_loss1, x='toss_winner', y='sum(win1)', title='toss_winner is not match_loser and toss_decition == bat')
      self.fig3.show()


      self.batting = self.df_matches.select(['toss_winner', 'toss_decision', 'winner']).filter((self.df_matches['toss_winner']==self.df_matches['winner']) & 
                                                                     (self.df_matches['toss_decision']=='bat')).select('toss_winner').groupBy('toss_winner').count().toPandas()                                                            
      self.fielding = self.df_matches.select(['toss_winner', 'toss_decision', 'winner']).filter((self.df_matches['toss_winner']==self.df_matches['winner']) & 
                                                                     (self.df_matches['toss_decision']=='field')).select('toss_winner').groupBy('toss_winner').count().toPandas()
      self.total_count = self.df_matches.select('toss_winner').groupBy('toss_winner').count().toPandas()     
      self.batting['percentage_winning'] = self.batting['count']/self.total_count['count']
      self.fielding['percentage_winning'] = self.fielding['count']/self.total_count['count'] 
      self.fig_final = px.bar(batting, x='toss_winner', y='percentage_winning', title = 'Percentage of team win where the toss decision is bat')
      self.fig_final.show() 
      self.fig_final1 = px.bar(fielding, x='toss_winner', y='percentage_winning', title = 'Percentage of team win where the toss decision is fielding')
      self.fig_final1.show()
      return {'Q10': 'With respect to toss decision there is a impact in result'}                                                   


    except:
      print('Something Wrong with the query10 code.  Please check and correct it, or try to run section 2 also')


In [ ]:
db = Database()
db.get_query1_result()

{'Q1': ['Dubai International Cricket Stadium',
  'Sheikh Zayed Stadium',
  'Sardar Patel Stadium, Motera']}